File: DSA_2215/sort/bubble/bubble_sort.c
#include <stdio.h>

enum SortType { ASCENDING, DESCENDING };

void bubbleSort(int arr[], int n, enum SortType sortType) {
  int swapped;
  do {
    swapped = 0;
    for (int i = 1; i < n; i++) {
      int shouldSwap = (sortType == ASCENDING) ? (arr[i - 1] > arr[i]) : (arr[i - 1] < arr[i]);
      if (shouldSwap) {
        int temp = arr[i - 1];
        arr[i - 1] = arr[i];
        arr[i] = temp;
        swapped = 1;
      }
    }
  } while (swapped);
}

int useMe() {
  int arr[] = {64, 25, 12, 22, 11};
  int n = sizeof(arr) / sizeof(arr[0]);
  enum SortType sortType = ASCENDING;  // Change to DESCENDING for descending

  bubbleSort(arr, n, sortType);

  printf("Sorted array: \n");
  for (int i = 0; i < n; i++) {
    printf("%d ", arr[i]);
  }
  return 0;
}

File: DSA_2215/sort/bubble/bubble_sort.dart
enum SortType { Ascending, Descending }

void bubbleSort(List<int> arr, SortType sortType) {
  int n = arr.length;
  for (int i = 0; i < n - 1; i++) {
    bool swapped = false;
    for (int j = 0; j < n - i - 1; j++) {
      bool shouldSwap =
          (sortType == SortType.Ascending && arr[j] > arr[j + 1]) ||
              (sortType == SortType.Descending && arr[j] < arr[j + 1]);
      if (shouldSwap) {
        int temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
        swapped = true;
      }
    }
    if (!swapped) {
      break;
    }
  }
}

void useme() {
  List<int> arr = [64, 25, 12, 22, 11];
  SortType sortType =
      SortType.Ascending; // Change to SortType.Descending for descending
  bubbleSort(arr, sortType);
  print("Sorted array: $arr");
}
